File: evals/ingest/traces.ts
import { appendFileSync, existsSync, mkdirSync, readFileSync } from 'node:fs'
import { isFromService, stripClientMetadata } from '../lib/ingest-filter.ts'

/**
 * Free-plan Workers Logs are retained 3 days, so a missed run loses those turns permanently.
 * This exports the whole window unfiltered rather than only the agent's structured `at` lines:
 * uncaught exceptions and request-level records are what a vanished turn leaves behind, and
 * deciding at export time that they are uninteresting is a decision that cannot be revisited.
 */

const ACCOUNT = process.env.CF_ACCOUNT_ID
const TOKEN = process.env.CF_API_TOKEN
const OUT = 'evals/data/traces.jsonl'
const SERVICE = 'personal-agent'
const PAGE = 2000
const MAX_PAGES = 50
const WINDOW_DAYS = Number.parseInt(process.env.TRACE_WINDOW_DAYS ?? '3', 10)

if (!ACCOUNT || !TOKEN) {
  console.error('set CF_ACCOUNT_ID and CF_API_TOKEN (Account Analytics: Read) in .env')
  process.exit(1)
}

type Event = { $metadata?: { id?: string } }

const seen = new Set<string>()
if (existsSync(OUT)) {
  for (const line of readFileSync(OUT, 'utf8').split('\n')) {
    if (!line.trim()) {
      continue
    }
    const id = (JSON.parse(line) as Event).$metadata?.id
    if (id) {
      seen.add(id)
    }
  }
}
const before = seen.size

const to = new Date()
const from = new Date(to.getTime() - WINDOW_DAYS * 24 * 60 * 60 * 1000)

async function page(offset?: string): Promise<Event[]> {
  const res = await fetch(
    `https://api.cloudflare.com/client/v4/accounts/${ACCOUNT}/workers/observability/telemetry/query`,
    {
      method: 'POST',
      headers: { authorization: `Bearer ${TOKEN}`, 'content-type': 'application/json' },
      body: JSON.stringify({
        queryId: 'trace-export',
        view: 'events',
        limit: PAGE,
        // Defaults to true server-side, which returns metadata and no events.
        dry: false,
        ...(offset ? { offset, offsetDirection: 'next' } : {}),
        parameters: { datasets: [] },
        // The API takes epoch millis here; ISO strings are rejected with a 400.
        timeframe: { from: from.getTime(), to: to.getTime() },
      }),
    },
  )
  const text = await res.text()
  if (!res.ok) {
    console.error(`observability query failed: ${res.status} ${text.slice(0, 500)}`)
    process.exit(1)
  }
  const body = JSON.parse(text) as { result?: { events?: { events?: Event[] } } }
  const events = body.result?.events?.events
  if (!events) {
    console.error(`unexpected response shape: ${text.slice(0, 500)}`)
    process.exit(1)
  }
  return events
}

mkdirSync('evals/data', { recursive: true })

let added = 0
let offset: string | undefined
for (let p = 0; p < MAX_PAGES; p++) {
  const events = await page(offset)
  if (events.length === 0) {
    break
  }

  let fresh = 0
  let foreign = 0
  for (const e of events) {
    const id = e.$metadata?.id
    if (!id || seen.has(id)) {
      continue
    }
    seen.add(id)
    // The account runs more than one Worker and the query returns all of them.
    if (!isFromService(e, SERVICE)) {
      foreign++
      continue
    }
    appendFileSync(OUT, `${JSON.stringify(stripClientMetadata(e))}\n`)
    added++
    fresh++
  }

  if (foreign) {
    console.log(`  page ${p}: skipped ${foreign} records from other workers`)
  }
  const last = events[events.length - 1]?.$metadata?.id
  // A page whose ids are all known still has to be walked past: the window is re-queried from
  // the start on every run, so early pages are duplicates by design and are not an end signal.
  if (events.length < PAGE || !last || last === offset) {
    break
  }
  offset = last
  if (fresh === 0 && p > 0 && events.length < PAGE) {
    break
  }
}

const total = existsSync(OUT) ? readFileSync(OUT, 'utf8').split('\n').filter(Boolean).length : 0
console.log(`${added} new records, ${total} in ${OUT} (was ${before})`)
if (added === 0 && before === 0) {
  console.error('nothing exported — check the token has Account Analytics: Read on this account')
  process.exit(1)
}

File: evals/ingest/traces.py
import json
import os
import time
from typing import Any

from lib.env import require_env
from lib.http import HttpError, dig, post_json
from lib.ingest_filter import is_from_service, strip_client_metadata
from lib.jsonl import append_json_line, read_json_lines
from lib.paths import DATA

OUT = DATA / 'traces.jsonl'
SERVICE = 'personal-agent'
PAGE = 2000
MAX_PAGES = 50
WINDOW_DAYS = int(os.environ.get('TRACE_WINDOW_DAYS', '3'))


def event_id(event: object) -> str | None:
    id = dig(event, '$metadata', 'id')
    return id if isinstance(id, str) else None


def page(account: str, token: str, from_ms: int, to_ms: int, offset: str | None) -> list[Any]:
    body: dict[str, Any] = {
        'queryId': 'trace-export',
        'view': 'events',
        'limit': PAGE,
        'dry': False,
        'parameters': {'datasets': []},
        'timeframe': {'from': from_ms, 'to': to_ms},
    }
    if offset:
        body['offset'] = offset
        body['offsetDirection'] = 'next'
    try:
        data = post_json(
            f'https://api.cloudflare.com/client/v4/accounts/{account}/workers/observability/telemetry/query',
            body,
            token,
        )
    except HttpError as err:
        raise SystemExit(f'observability query failed: {err.status} {err.body[:500]}') from err
    events: list[Any] | None = dig(data, 'result', 'events', 'events')
    if events is None:
        raise SystemExit(f'unexpected response shape: {json.dumps(data)[:500]}')
    return events


def main() -> None:
    account, token = require_env('CF_ACCOUNT_ID', 'CF_API_TOKEN')
    seen = {id for id in (event_id(e) for e in read_json_lines(OUT)) if id}
    before = len(seen)
    to_ms = int(time.time() * 1000)
    from_ms = to_ms - WINDOW_DAYS * 24 * 60 * 60 * 1000

    added = 0
    offset: str | None = None
    for p in range(MAX_PAGES):
        events = page(account, token, from_ms, to_ms, offset)
        if not events:
            break
        fresh = 0
        foreign = 0
        for e in events:
            id = event_id(e)
            if not id or id in seen:
                continue
            seen.add(id)
            if not is_from_service(e, SERVICE):
                foreign += 1
                continue
            append_json_line(OUT, strip_client_metadata(e))
            added += 1
            fresh += 1
        if foreign:
            print(f'  page {p}: skipped {foreign} records from other workers')
        last = event_id(events[-1])
        if len(events) < PAGE or not last or last == offset:
            break
        offset = last

    total = len(read_json_lines(OUT))
    print(f'{added} new records, {total} in {OUT} (was {before})')
    if added == 0 and before == 0:
        raise SystemExit('nothing exported — check the token has Account Analytics: Read on this account')


if __name__ == '__main__':
    main()
